### PR TITLE
Prevent sh_malloc infinite loop

### PR DESCRIPTION
**Pull request by guidovranken**
##### Description of change
This prevents ```sh_malloc``` to stall in an infinite for loop. Specifically, if the size requested is larger than the largest multiple of ```sh.minsize``` before being reset to 0 (due to left-shift overflow), the loop will go on forever. From a security perspective this is especially unfavorable behavior because excessive allocations that would otherwise (with traditional malloc) would fail silently, turn into remote denial-of-service vulnerabilities due to this bug. This PR fixes this issue.

### CODE REVIEW

**Comment by guidovranken** [comment]
Found another infinite loop, added fix and test to this PR as well.

File: crypto/mem_sec.c [hunk 1]
```diff
@@ -476,8 +480,11 @@ static char *sh_malloc(size_t size)
     char *chunk;
 
     list = sh.freelist_size - 1;
-    for (i = sh.minsize; i < size; i <<= 1)
+    for (i = sh.minsize; i && i < size; i <<= 1)
```
Comment (levitte) [inline crypto/mem_sec.c]: For variables that aren't meant to be boolean, we prefer explicit tests, i.e.:

``` C
    for (i = sh.minsize; i > 0 && i < size; i <<= 1)
```

Comment (guidovranken) [inline crypto/mem_sec.c]: Thanks I will keep that in mind for next time.

Comment (levitte) [inline crypto/mem_sec.c]: Note that it's a change I request that you do

**Comment by richsalz** [comment]
can you squash/merge and push an update so it's a single commit?

**Comment by richsalz** [comment]
Argh, conflicts.  Can you rebase against master?  thanks.

**Comment by guidovranken** [comment]
Will this work?

**Comment by guidovranken** [comment]
I don't understand. Please tell me what to do.

**Comment by richsalz** [comment]
In your workspace:
```
git co secheap_infinite_loop_fix
git rebase master
git push -f github HEAD
```

**Comment by richsalz** [comment]
Don;t worry about the squash/merge, i can handle that.
Now, is this needed?  Or is it fixed by #2657 ?

**Comment by richsalz** [comment]
I took off one reviewer, pending feedback from @guidovranken

**Comment by guidovranken** [comment]
Once the RSA_padding_add_PKCS1_PSS_mgf1 PR is closed I am going to remove my fork of openssl and re-fork it. With my first pull request I made the mistake to apply the changes to my own master instead of creating a new branch for it, and each time I do rebase master now, I get conflicts.. I will create a new PR for this.

File: crypto/mem_sec.c [hunk 2]
```diff
@@ -356,6 +357,8 @@ static int sh_init(size_t size, int minsize)
     sh.minsize = minsize;
     sh.bittable_size = (sh.arena_size / sh.minsize) * 2;
 
+    if (sh.bittable_size > INT_MAX)
+      goto err;
```
Comment (TheMassController) [inline crypto/mem_sec.c]: Why is this necessary?

Comment (richsalz) [inline crypto/mem_sec.c]: because bittable_size is a size_t, unsigned int, and other parts of the code treat dependant data structures as if things are int-sized.

**Comment by richsalz** [comment]
ping.  This is a bugfix :)

**Comment by richsalz** [comment]
AARGH.  Needs to be rebased (although I can address the conflict), but one test fails now.  Ping @guidovranken sorry for taking so long.

File: test/secmemtest.c [hunk 3]
```diff
@@ -90,6 +96,11 @@ int main(int argc, char **argv)
         perror_line();
         return 1;
     }
+    /* Test for infinite loop bug in sh_init, should fail */
+    if (CRYPTO_secure_malloc_init(1 << 30, 1)) {
```
Comment (levitte) [inline test/secmemtest.c]: Why do you think this will fail?  Sure, on a machine with 32-bit `size_t`, but on a machine with 64-bit, you're nog even coming near `INT_MAX`.  I suggest this instead:

``` C
    if (CRYPTO_secure_malloc_init(1 << (sizeof(size_t) * 8 - 2), 1)) {
```

**Comment by richsalz** [comment]
@guidovranken can you refresh /rebase /update this?

**Comment by guidovranken** [comment]
Is this alright?

**Comment by richsalz** [comment]
@levitte please take a look now.

**Comment by richsalz** [comment]
Sorry, after all that it was done in commit c181fdbe3c52a10ff6203c3219c57fbbddb599a3 in February 2017!  Sheesh!
